test/catch.cc

```diff
@@ -0,0 +1,215 @@
+#include "eventuals/catch.h"
+
+#include <iostream>
+#include <string>
+
+#include "eventuals/eventual.h"
+#include "eventuals/expected.h"
+#include "eventuals/just.h"
+#include "eventuals/raise.h"
+#include "eventuals/terminal.h"
+#include "eventuals/then.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using eventuals::Catch;
+using eventuals::Eventual;
+using eventuals::Expected;
+using eventuals::Just;
+using eventuals::Raise;
+using eventuals::Then;
+using eventuals::Unexpected;
+using testing::MockFunction;
+
+TEST(CatchTest, RaisedRuntimeError) {
+  auto e = []() {
+    return Just(1)
+        | Raise(std::runtime_error("message"))
+        | Catch()
+              .raised<int>([](auto&& error) {
+                ADD_FAILURE() << "Encountered an unexpected raised 'int'";
+                return 1;
+              })
+              .raised<std::runtime_error>([](auto&&) {
+                return 100;
+              });
+  };
+
+  EXPECT_EQ(*e(), 100);
+}
+
+TEST(CatchTest, ChildException) {
+  struct Error : public std::exception {
+    const char* what() const noexcept override {
+      return "child exception";
+    }
+  };
+
+  auto e = [&]() {
+    return Just(1)
+        | Raise(Error{})
+        | Catch()
+              .raised<int>([](auto&& error) {
+                ADD_FAILURE() << "Encountered an unexpected raised 'int'";
+                return Just(10);
+              })
+              .raised<std::exception>(
+                  [](auto&& error) {
+                    EXPECT_EQ(
+                        std::string("child exception"),
+                        error.what());
+                    return Just(100);
+                  });
+  };
+
+  EXPECT_EQ(*e(), 100);
+}
+
+TEST(CatchTest, All) {
+  auto checker = [](auto error) {
+    EXPECT_EQ(error, 10);
+  };
+
+  auto e = [&]() {
```

Comment:
This is exactly why we want to have `Fail(Error&& error)` functions because then we don't have to jump through hoops like this because we have a `...`. It would be much better to just check inline, it's less complex!
```suggestion
  auto e = []() {
```